### PR TITLE
Implemented pagination

### PR DESCRIPTION
**Pull request by cyberdr0id**

### CODE REVIEW

File: internal/handler/handlers.go [hunk 1]
```diff
@@ -158,13 +159,24 @@ func (s *Server) GetRequests(rw http.ResponseWriter, r *http.Request) {
...
+	}
+	if err != nil {
+		sendResponse(rw, err.Error(), http.StatusInternalServerError)
+		return
+	}
 
 	if err := ValidateRequestState(t); err != nil {
 		sendResponse(rw, err.Error(), http.StatusBadRequest)
 		return
 	}
```
Comment (papahmsolo) [inline internal/handler/handlers.go:177]: Single validate function

File: internal/handler/handlers.go [hunk 2]
```diff
@@ -319,15 +331,15 @@ func ValidateRequestState(state string) error {
...
 	if !ok {
-		return ok, fmt.Errorf("%w: id has bad format", ErrInvalidParameter)
+		return ok, fmt.Errorf("%w: numeric parameter has bad format", ErrInvalidParameter)
 	}
 	if err != nil {
-		return ok, fmt.Errorf("cannot validate user ID: %w", err)
+		return ok, fmt.Errorf("cannot validate input parameter: %w", err)
 	}
 
 	return ok, nil
```
Comment (papahmsolo) [inline internal/handler/handlers.go:345]: returns error only

**Review by papahmsolo — APPROVED** [review]
👀

File: internal/service/referral.go [hunk 3]
```diff
@@ -59,13 +60,23 @@ func (s *ReferralService) AddCandidate(ctx context.Context, request SubmitCandid
...
-	requests, err := s.repo.GetRequests(userID, status)
+	pageNumberInt, err := strconv.Atoi(pageNumber)
+	if err != nil {
+		return nil, fmt.Errorf("cannot convert page number to integer: %w", err)
+	}
+
+	pageSizeInt, err := strconv.Atoi(pageSize)
+	if err != nil {
+		return nil, fmt.Errorf("cannot convert page size to integer: %w", err)
+	}
```
Comment (papahmsolo) [inline internal/service/referral.go:77]: Already explained in the previous comment

File: internal/repository/referral.go [hunk 4]
```diff
@@ -16,16 +16,33 @@ type UserRequests struct {
...
-				 FROM requests WHERE author_id = %s`, id)
+		query = fmt.Sprintf(`
+				SELECT 
+					id, candidate_name, candidate_surname, status, updated 
+				FROM 
+					requests 
+				WHERE author_id = %s
+				LIMIT %d
+				OFFSET %d
+				`, id, pageSize, offset)
```
Comment (papahmsolo) [inline internal/repository/referral.go:34]: We're not validating and escaping input. Potential SQL injection ([more info](https://go.dev/doc/database/sql-injection)). Have you run a linter here?

File: internal/repository/referral.go [hunk 5]
```diff
@@ -16,16 +16,33 @@ type UserRequests struct {
...
+		query = fmt.Sprintf(`
+				SELECT 
+					id, candidate_name, candidate_surname, status, updated 
+				FROM 
+					requests 
+				WHERE 
+					author_id = %s AND requests.status = '%s'
+				LIMIT %d
+				OFFSET %d
+				`, id, status, pageSize, offset)
```
Comment (papahmsolo) [inline internal/repository/referral.go:45]: This is almost identical to previous query, if you can find a way to minimize code duplication that'd be great. If not, that's fine, leave it as is.

**Review by papahmsolo — CHANGES_REQUESTED** [review]
👀

File: internal/repository/referral.go (hunk 4 above)
Comment (cyberdr0id) [reply internal/repository/referral.go:34]: yep, but nothing happened

File: internal/repository/referral.go [hunk 6]
```diff
@@ -16,23 +17,43 @@ type UserRequests struct {
...
+	whereVal = append(whereVal, pageSize, offset)
+
+	query = fmt.Sprintf(`
+			SELECT 
+				id, candidate_name, candidate_surname, status, updated 
+			FROM 
+				requests 
+			WHERE `+strings.Join(whereCol, " AND ")+` LIMIT $%d OFFSET $%d`, paramNumber, paramNumber+1)
+
+	rows, err := r.db.Query(query, whereVal...)
```
Comment (papahmsolo) [inline internal/repository/referral.go:53]: queries with and without status are very similar

File: internal/handler/handlers.go [hunk 7]
```diff
@@ -158,13 +164,35 @@ func (s *Server) GetRequests(rw http.ResponseWriter, r *http.Request) {
...
+	if err != nil {
+		sendResponse(rw, fmt.Errorf("cannot convert page number to integer: %w", err), http.StatusBadRequest)
+		return
+	}
+
+	pageSizeInt, err := strconv.Atoi(pageSize)
+	if err != nil {
+		sendResponse(rw, fmt.Errorf("cannot convert page size to integer: %w", err), http.StatusBadRequest)
+		return
+	}
```
Comment (papahmsolo) [inline internal/handler/handlers.go:193]: single validation func

**Review by papahmsolo — CHANGES_REQUESTED** [review]
👀

**Review by papahmsolo — APPROVED** [review]

**Pull request merged**